typescript/sdk/src/ism/metadata/builder.ts

```diff
@@ -0,0 +1,139 @@
+  Address,
+  ParsedMessage,
+  WithAddress,
+  assert,
+  rootLogger,
+} from '@hyperlane-xyz/utils';
+
+import { deepFind } from '../../../../utils/dist/objects.js';
+import { HyperlaneCore } from '../../core/HyperlaneCore.js';
+import { DispatchedMessage } from '../../core/types.js';
+import { DerivedHookConfig } from '../../hook/EvmHookReader.js';
+import { HookType, MerkleTreeHookConfig } from '../../hook/types.js';
+import { MultiProvider } from '../../providers/MultiProvider.js';
+import { DerivedIsmConfig } from '../EvmIsmReader.js';
+import { IsmConfig, IsmType } from '../types.js';
+
+import {
+  AggregationMetadata,
+  AggregationMetadataBuilder,
+} from './aggregation.js';
+import { MultisigMetadata, MultisigMetadataBuilder } from './multisig.js';
+import { NullMetadata, NullMetadataBuilder } from './null.js';
+import { RoutingMetadata, RoutingMetadataBuilder } from './routing.js';
+
+export type StructuredMetadata =
+  | NullMetadata
+  | MultisigMetadata
+  | AggregationMetadata<any>
+  | RoutingMetadata<any>;
+
+export interface MetadataContext<
+  IsmContext = DerivedIsmConfig,
+  HookContext = DerivedHookConfig,
+> {
+  message: DispatchedMessage;
+  dispatchTx: TransactionReceipt;
+  ism: IsmContext;
+  hook: HookContext;
+}
+
+export interface MetadataBuilder {
+  build(context: MetadataContext): Promise<string>;
+}
+
+export class BaseMetadataBuilder implements MetadataBuilder {
+  public nullMetadataBuilder: NullMetadataBuilder;
+  public multisigMetadataBuilder: MultisigMetadataBuilder;
+  public aggregationMetadataBuilder: AggregationMetadataBuilder;
+  public routingMetadataBuilder: RoutingMetadataBuilder;
+
+  public multiProvider: MultiProvider;
+  protected logger = rootLogger.child({ module: 'BaseMetadataBuilder' });
+
+  constructor(core: HyperlaneCore) {
+    this.multisigMetadataBuilder = new MultisigMetadataBuilder(core);
+    this.aggregationMetadataBuilder = new AggregationMetadataBuilder(this);
+    this.routingMetadataBuilder = new RoutingMetadataBuilder(this);
+    this.nullMetadataBuilder = new NullMetadataBuilder(core.multiProvider);
+    this.multiProvider = core.multiProvider;
+  }
+
+  // assumes that all post dispatch hooks are included in dispatchTx logs
+  async build(context: MetadataContext, depth = 10): Promise<string> {
+    this.logger.debug(
+      { context, depth },
+      `Building ${context.ism.type} metadata`,
+    );
+    assert(depth > 0, 'Max depth reached');
+
+    const { ism, hook } = context;
+    switch (ism.type) {
+      case IsmType.TRUSTED_RELAYER:
+      case IsmType.TEST_ISM:
+      case IsmType.OP_STACK:
+      case IsmType.PAUSABLE:
+        return this.nullMetadataBuilder.build({ ...context, ism });
+
+      case IsmType.MERKLE_ROOT_MULTISIG:
+      case IsmType.MESSAGE_ID_MULTISIG:
+        const merkleTreeHook = deepFind(
+          hook,
+          (v): v is WithAddress<MerkleTreeHookConfig> =>
+            v.type === HookType.MERKLE_TREE && v.address !== undefined,
```

Comment:
`&& !!v.address` is sufficient

Reply:
ngl I find this a little less readable

Reply:
Fair enough. What's the expected behavior if address is `null` or `''`? Because the !! cover all 3 whereas this does not

Reply:
hmm good point
we shouldnt return true if its null or ''